### PR TITLE
Bluetooth: Mesh: Access model callback struct

### PR DESCRIPTION
**Pull request by trond-snekvik**
Adds a structure of callbacks for each model instance. This allows for more flexible model implementations, that can interact with the Mesh stack without going through the application.

The motivation for this change is to improve the user experience when adding modular, reusable model implementations. The intention is to make it as easy to mix and match models as it is with drivers, where the user only has to list their models in the composition data to enable them. 

Added callbacks for init and model user data persistent storage. The init callback replaces the internal lookup for foundation models. Also adds variation of the `BT_MESH_MODEL` macro with a callback parameter, without breaking the backward compatibility of the old macro.

Looking for some input, particularly in these areas:
Any other callbacks we should add right away? 
Are there better ways to enable model instance settings storage? 
Should the model persistent storage be scheduled like the other mesh settings?

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]

**Comment by jhedberg** [comment]
@trond-snekvik looks good to me, but there seems to have crept in some conflict with `access.h` - you'll need to rebase against latest master and fix it before I can merge this.

**Comment by jhedberg** [comment]
One thing that I started wondering about: what about node reset? How are you planning to handle that? Should the respective model data be deleted from flash in that case, or do you just leave it hanging around there until the node gets provisioned again? Should there be some new callback for this. We do have a global reset callback currently though.

**Comment by trond-snekvik** [comment]
That's a good point. I think we should delete the model data on reset, and as a consequence, each model should probably be notified of this change. I can add this to this PR.

File: subsys/bluetooth/mesh/cfg_srv.c [hunk 1]
```diff
@@ -3273,21 +3278,33 @@ int bt_mesh_cfg_srv_init(struct bt_mesh_model *model, bool primary)
...
 	clear_count = mod_sub_list_clear(mod);
 
-	if (IS_ENABLED(CONFIG_BT_SETTINGS) && clear_count) {
-		bt_mesh_store_mod_sub(mod);
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		if (clear_count) {
+			bt_mesh_store_mod_sub(mod);
+		}
+
+		bt_mesh_model_data_store(mod, vnd, NULL, 0);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/cfg_srv.c]: If I'm reading the settings module correctly, this'll actually generate 0-length entries for models that don't have any user data. I couldn't find a good way to get around this though, do you have any suggestions, or do you think it's fine @jhedberg?

Comment (jhedberg) [reply subsys/bluetooth/mesh/cfg_srv.c]: A new flag to `model->flags` to indicate if data was ever stored to or retrieved from flash? Similar to what I've tried to do here for pairing keys: https://github.com/jhedberg/zephyr/commit/7e22ca25bc53a97331876bff97c232dcc20a0333

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/cfg_srv.c]: Of course 🤦‍♂ Made the change.

**Pull request merged**